pkg/clustersserver/clustersserver.go

```diff
@@ -28,21 +30,44 @@ type Server struct {
 	ClientCache       clientCache
 	AvailableContexts []string
 	InitialContext    string
+	CreateClient      func(string) (client.Client, error)
+	mu                sync.Mutex
+}
+
+func NewServer(kubeContexts []string, currentKubeContext string) http.Handler {
+	clusters := Server{
+		AvailableContexts: kubeContexts,
+		InitialContext:    currentKubeContext,
+		ClientCache:       map[string]client.Client{},
+		CreateClient:      defaultCreateClient,
+	}
+
+	clustersHandler := pb.NewClustersServer(&clusters, nil)
+
+	return clustersHandler
+}
+
+func defaultCreateClient(kubeContext string) (client.Client, error) {
+	return util.NewKubeClient(kubeContext)
 }
 
 func (s *Server) getClient(kubeContext string) (client.Client, error) {
 	if s.ClientCache[kubeContext] != nil {
 		return s.ClientCache[kubeContext], nil
 	}
 
-	client, err := util.NewKubeClient(kubeContext)
+	s.mu.Lock()
```

Comment:
You need to do the read while holding the lock, too. As an example of why this matters: you could read `s.ClientCache[kubeContext]`, then another process could delete it before you return `s.ClientCache[kubeContext]` in the next line.

You could do

```
if client, ok := s.ClientCache[kubeContext]; ok {
    return client, nil
}
```

... so you only read once. But, now you have the problem of the other branch -- what if there's no value when you look, but someone puts one there after? You will needlessly create a new client and overwrite the other.

You should also make sure that `s.mu.Unlock()` is _always_ called, either by putting it before each return from the procedure, or (more conveniently here) using `defer s.mu.Unlock()` right after you obtain the lock.